pkg/addons/addons.go

```diff
@@ -85,6 +88,83 @@ func readDB(cs *acsapi.ContainerService, c *config.Config) (map[string]unstructu
 	return db, nil
 }
 
+// syncWorkloadsConfig iterates over all workload controllers (deployments,
+// daemonsets, statefulsets), walks their volumes, and updates their pod
+// templates with annotations that include the hashes of the content for
+// each configmap or secret.
+func syncWorkloadsConfig(db map[string]unstructured.Unstructured) error {
+	configToHash := make(map[string]string)
+	for _, o := range db {
+		if o.GroupVersionKind().Kind != "Secret" &&
+			o.GroupVersionKind().Kind != "ConfigMap" {
+			continue
+		}
+
+		for _, v := range jsonpath.MustCompile("$.data").Get(o.Object) {
```

Comment:
>  1- this looks wrong: hasher should be instantiated outside the loops and configToHash should be written once at the end.

The loop really only runs once since there is only a .data or .stringData field in every object. Techincally there is no difference from what you are proposing but I can change to it.

>  2- I only see potential downside in using fnv - suggest sha256 to match helm.

ack

> ignore this one if you want: the secret or configmap is "cleaned" at this point; assuming a serialiser is used which sorts keys it should be possible simply to hash the serialised object and not iterate its fields. This would be useful if we generalise this beyond secrets and configmaps.

I prefer not to depend on the "cleaning" of the objects and be on the safe side with hashing just the config content.

Reply:
> This would be useful if we generalise this beyond secrets and configmaps.

Do you have anything specific in mind apart from CMs/secrets where we could use this?

Reply:
> The loop really only runs once since there is only a .data or .stringData field in every object. Techincally there is no difference from what you are proposing but I can change to it.

ah, sorry, I thought it was running for `$.data.*`.  Since it isn't, now I'm wondering whether the iteration order, and hence the hash, is stable?

> I prefer not to depend on the "cleaning" of the objects 

Understood, although we're somewhat reliant on this elsewhere already.

> Do you have anything specific in mind apart from CMs/secrets where we could use this?

I haven't thought in detail how far this can be generalised, but I wonder whether it can be done for things like service account configuration as well, for example (but not in this PR).

Reply:
> ah, sorry, I thought it was running for $.data.*. Since it isn't, now I'm wondering whether the iteration order, and hence the hash, is stable?

By iteration order you mean the order objects are iterated? Can't see what that has to do with hashing, mind elaborating?

Reply:
the hash of [a, b] is different to the hash of [b, a], and the iteration order of Go maps is undefined and deliberately randomized.  But it turns out that %#v sorts keys as it iterates them.

Reply:
the iteration order must be stable else the hash would change all the time and we'd redeploy all the time.